tool_path_planner/src/raster_path_generator.cpp

```diff
@@ -0,0 +1,337 @@
+  // run along the X direction and are spaced out ~ in Y
+  auto end_points = toEndPoints(eigen_poses, avg_quaternion);
+
+  // Sort end points, -y to y
+  std::sort(end_points.begin(), end_points.end(), [] (const PathEndPoints& lhs, const PathEndPoints& rhs)
+  {
+    auto lhs_value = std::min(lhs.a.y(), lhs.b.y());
+    auto rhs_value = std::min(rhs.a.y(), rhs.b.y());
+    return lhs_value < rhs_value;
+  });
+
+  // A helper function to get the starting point of a transition given a sequence number and
+  // whether we started at A or B.
+  auto current_position = [&end_points](const SequencePoint& p) {
+    if (p.from_a) // If we came from A, we're now at B
+      return end_points[p.id].b;
+    else // if we came from B, we're not at A
+      return end_points[p.id].a;
+  };
+
+  std::vector<SequencePoint> sequence;
+  sequence.reserve(input.size());
+
+  // We always start at the first end_point, position A
+  sequence.push_back({0, true});
+
+  for (std::size_t i = 1; i < end_points.size(); ++i)
+  {
+    // We need to determine if A or B of the next path is closer to the current position
+    const Eigen::Vector3d current_pos = current_position(sequence.back());
+
+    const auto dist_a = (end_points[i].a - current_pos).squaredNorm();
+    const auto dist_b = (end_points[i].b - current_pos).squaredNorm();
+
+    const auto from_a = dist_a < dist_b;
+    sequence.push_back({i, from_a});
+  }
+
+  // Re-order the original inputs and produce a new segment.
+  return makeSequence(input, sequence, end_points);
+}
+
+
+RasterPathGenerator::RasterPathGenerator()
+{
+
+}
+
+RasterPathGenerator::~RasterPathGenerator()
+{
+
+}
+
+tool_path_planner::ProcessTool RasterPathGenerator::generateDeaultToolConfig()
+{
+  tool_path_planner::ProcessTool tool;
+  tool.pt_spacing = 0.01;
+  tool.line_spacing = 0.015;
+  tool.tool_offset = 0.0; // currently unused
+  tool.intersecting_plane_height = 0.5; // 0.5 works best, not sure if this should be included in the tool
+  tool.min_hole_size = 0.01;
+  return tool;
+}
+
+boost::optional< std::vector<geometry_msgs::PoseArray> > RasterPathGenerator::generate(const tool_path_planner::ProcessTool& path_gen_config,
+                                                                      const pcl::PolygonMesh& mesh) const
+{
+  //
+  using ToolPath = std::vector<tool_path_planner::ProcessPath>;
+
+  // instantiate raster tool planner
+  tool_path_planner::RasterToolPathPlanner tool_path_planner;
+  tool_path_planner.setTool(path_gen_config);
+  std::vector<ToolPath> tool_process_paths;
+  tool_path_planner.planPaths({mesh}, tool_process_paths);
+  if(tool_process_paths.size() != 1)
+  {
+    return  boost::none;
+  }
+
+  bool all_empty = std::all_of(tool_process_paths.begin(),tool_process_paths.end(),[](auto& tp){
+    return tp.empty();
+  });
+  if(all_empty)
+  {
+    ROS_ERROR("All tool paths generated are empty");
+    return boost::none;
+  }
+
+  ROS_INFO("Found %lu tool paths",tool_process_paths.size());
```

Comment:
It seems odd to check all of them to see if they are all empty, then to say we have found n paths when many might be empty. The should probably count the non-empty tool paths